src/Raven.Server/Smuggler/Migration/Migrator.cs

```diff
@@ -0,0 +1,353 @@
+        {
+            var operationId = _operations.GetNextOperationId();
+            var cancelToken = new OperationCancelToken(_cancellationToken);
+            var result = new SmugglerResult();
+
+            _operations.AddOperation(null,
+                $"Migrating database: {sourceDatabaseName} from url: {_serverUrl}",
+                Operations.OperationType.DatabaseImport,
+                taskFactory: onProgress => Task.Run(async () =>
+                {
+                    onProgress?.Invoke(result.Progress);
+
+                    using (cancelToken)
+                    {
+                        try
+                        {
+                            await DoMigration(sourceDatabaseName, result, onProgress, database, cancelToken);
+                        }
+                        catch (Exception e)
+                        {
+                            result.AddError($"Error occurred during database migration named: {sourceDatabaseName}." +
+                                            $"Exception: {e.Message}");
+
+                            throw;
+                        }
+                    }
+
+                    return (IOperationResult)result;
+                }, cancelToken.Token), id: operationId, token: cancelToken);
+
+            return operationId;
+        }
+
+        private async Task DoMigration(
+            string databaseName, 
+            SmugglerResult result, 
+            Action<IOperationProgress> onProgress,
+            DocumentDatabase database,
+            OperationCancelToken cancelToken)
+        {
+            var state = GetLastMigrationState(database);
+            const ItemType types = ItemType.Documents | ItemType.Indexes | ItemType.Transformers;
+            //TODO: ItemType.Attachments;
+            var smugglerDatabaseOptions = new SmugglerDatabaseOptions
+            {
+                BatchSize = 1024,
+                OperateOnTypes = types,
+                //TODO: ExportDeletions = state != null,
+                StartDocsEtag = state?.LastDocsEtag ?? LastEtagsInfo.EtagEmpty,
+                StartDocsDeletionEtag = state?.LastDocDeleteEtag ?? LastEtagsInfo.EtagEmpty,
+                StartAttachmentsEtag = state?.LastAttachmentsEtag ?? LastEtagsInfo.EtagEmpty,
+                StartAttachmentsDeletionEtag = state?.LastAttachmentsDeleteEtag ?? LastEtagsInfo.EtagEmpty
+            };
+
+            var operationId = await GetOperationId(databaseName);
+            var exportData = new ExportData
+            {
+                DownloadOptions = JsonConvert.SerializeObject(smugglerDatabaseOptions),
+                ProgressTaskId = operationId
+            };
+
+            await MigrateDatabase(databaseName, exportData, database, result, onProgress, cancelToken);
+
+            await SaveLastState(databaseName, operationId, database);
+        }
+
+        private async Task SaveLastState(string databaseName, long operationId, DocumentDatabase database)
```

Comment:
what is this for?

Reply:
Saving the last import status from the destination